### PR TITLE
Add continued callbacks to scheduler

### PR DESCRIPTION
**Pull request by flarnie**
**Problem:**

We want to use JS Scheduler to coordinate non-React JS with React, so that we don't get low priority background JS work blocking a React render, like so - 
screen shot 2018-07-25 at 1 17 30 pm

Right now, even if the "lazy load JS" callback calls 'scheduleWork', it will only get pushed forward one frame, and could still run before React finishes high priority render work. The current implementation just does a naive FIFO ordering unless a callback reaches it's expiration time.

We want React's callback to keep getting run over and over until it's done, and *then* run any other work, like so - 
screen shot 2018-07-25 at 1 18 50 pm

**Initial Solution:**
@acdlite and I discussed adding some API to allow a "continued" callback, so that the scheduler knows to put the next segment of work for that callback at the head of the list rather than the tail.

I'm open to changing the API for this, adding a new option just seemed like the least invasive way, since adding a new method to the JS Scheduler would mean potentially injecting a new method into the renderers.

**More Solutions:**
I see a need for the JS Scheduler to know about different priorities in the future; any callback should be given a priority, and the priorities should mirror the ones used internally by React, that we've discussed before. I can go into why we need this if it's not clear.

I'm not sure that we will still need/want the `continued` option after adding support for priorities, but think we might. React's single recurring callback does seem like it should be treated as one continued unit of work.

**TODO next:**
I want to add an integration test, simulating the situation from the above diagrams. Wondering what directory would be best for adding a test of actually using the JS Scheduler and React together. And if there might be something simpler than copying all of the timer mocks from the current JS Scheduler test.

### CODE REVIEW

**Comment by react-linter** [comment]
ReactDOM: size: :small_red_triangle:+0.1%, gzip: :small_red_triangle:+0.1%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 0154a79fedef38a824a837c535bc853013dd4588...07d171ee3a959325f51a1e5cdd101394454f145d</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.1% | +0.1% | 641.25 KB | 641.92 KB | 150.38 KB | 150.5 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **96.08 KB** | **96.18 KB** | **31.09 KB** | **31.13 KB** | **UMD_PROD**
react-dom.development.js | +0.1% | +0.1% | 637.38 KB | 638.06 KB | 149.2 KB | 149.33 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **96.07 KB** | **96.17 KB** | **30.61 KB** | **30.66 KB** | **NODE_PROD**
ReactDOM-dev.js | 0.0% | 0.0% | 644.68 KB | 644.8 KB | 147.58 KB | 147.62 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **0.0%** | **:small_red_triangle:+0.1%** | **278.74 KB** | **278.82 KB** | **52.13 KB** | **52.16 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | +0.1% | +0.1% | 97.2 KB | 97.3 KB | 31.02 KB | 31.06 KB | NODE_PROFILING
ReactDOM-profiling.js | 0.0% | +0.1% | 282.04 KB | 282.12 KB | 52.85 KB | 52.88 KB | FB_WWW_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.2% | +0.1% | 427.78 KB | 428.45 KB | 96.57 KB | 96.69 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.2%** | **83.48 KB** | **83.58 KB** | **25.79 KB** | **25.84 KB** | **UMD_PROD**
react-art.development.js | +0.2% | +0.1% | 360.3 KB | 360.97 KB | 79.5 KB | 79.61 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.2%** | **48.47 KB** | **48.57 KB** | **15.18 KB** | **15.21 KB** | **NODE_PROD**
ReactART-dev.js | 0.0% | +0.1% | 350.55 KB | 350.66 KB | 74.38 KB | 74.42 KB | FB_WWW_DEV
**ReactART-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.1%** | **150.15 KB** | **150.23 KB** | **25.47 KB** | **25.5 KB** | **FB_WWW_PROD**

## react-scheduler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-scheduler.development.js | +3.7% | +1.6% | 15.39 KB | 15.96 KB | 4.64 KB | 4.71 KB | UMD_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+2.0%** | **:small_red_triangle:+1.9%** | **2.72 KB** | **2.77 KB** | **1.25 KB** | **1.28 KB** | **UMD_PROD**
react-scheduler.development.js | +3.7% | +1.6% | 15.2 KB | 15.77 KB | 4.6 KB | 4.67 KB | NODE_DEV
**react-scheduler.production.min.js** | **:small_red_triangle:+2.0%** | **:small_red_triangle:+1.9%** | **2.64 KB** | **2.69 KB** | **1.2 KB** | **1.22 KB** | **NODE_PROD**
ReactScheduler-dev.js | +3.7% | +1.6% | 15.42 KB | 15.98 KB | 4.64 KB | 4.71 KB | FB_WWW_DEV
**ReactScheduler-prod.js** | **:small_red_triangle:+3.5%** | **:small_red_triangle:+1.7%** | **7.58 KB** | **7.84 KB** | **1.89 KB** | **1.93 KB** | **FB_WWW_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by acdlite** [comment]
The most natural API for this would be a generator:

```js
scheduleWork(function *(deadline) {
  while (hasWork()) {
    if (deadline.timeRemaining() < 1) {
      // Yielding indicates there's still work left over
      yield;
    }
    doUnitOfWork();
  }
  // Returning indicates the work is complete.
});
```

but I wonder if the extra overhead is worth the convenience, since it's mostly only frameworks that will use this API.

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 1]
```diff
@@ -1920,7 +1924,9 @@ function performWork(minExpirationTime: ExpirationTime, dl: Deadline | null) {
   }
   // If there's work left over, schedule a new callback.
   if (nextFlushedExpirationTime !== NoWork) {
-    scheduleCallbackWithExpirationTime(nextFlushedExpirationTime);
+    scheduleCallbackWithExpirationTime(nextFlushedExpirationTime, {
+      continue: true,
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: `continue` should only be true if we were already inside an async callback. That's the check above: `deadline !== null`. Maybe scheduler should warn or error if `continue: true` is passed but you're not already inside a callback.

To me this indicates that the API should use generators instead. Then this type of mistake wouldn't be possible.

File: packages/react-scheduler/src/ReactScheduler.js [hunk 2]
```diff
@@ -375,14 +375,26 @@ if (!canUseDOM) {
       headOfPendingCallbacksLinkedList = scheduledCallbackConfig;
       tailOfPendingCallbacksLinkedList = scheduledCallbackConfig;
     } else {
-      // Add latest callback as the new tail of the list
-      scheduledCallbackConfig.prev = tailOfPendingCallbacksLinkedList;
-      // renaming for clarity
-      const oldTailOfPendingCallbacksLinkedList = tailOfPendingCallbacksLinkedList;
-      if (oldTailOfPendingCallbacksLinkedList !== null) {
-        oldTailOfPendingCallbacksLinkedList.next = scheduledCallbackConfig;
+      if (options && options.continued) {
```
Comment (acdlite) [inline packages/react-scheduler/src/ReactScheduler.js]: See other comment. What should happen if `options.continued` is true but we're not inside another callback?

**Comment by flarnie** [comment]
Agreed that generators would be a good fit for this use case, was thinking the lack of support in IE11 would block us using them here though.

> I wonder if the extra overhead is worth the convenience, since it's mostly only frameworks that will use this API.

Yea, that's why I wasn't too concerned with the API, but I like the idea of warning if 'continue' is passed and we are not already processing a callback. Also would like to enforce that only React uses this API, at least initially, but that also seems hard to enforce, didn't want to add complexity.

**Comment by flarnie** [comment]
Thinking about whether we could write a generator-able, to avoide the browser support issue. Even if that works I can think of some potential problems though, mainly that it's more lines of code.

**Comment by flarnie** [comment]
Playing around with this, I also wonder how JS Scheduler would know to stop calling the generator.

**Comment by acdlite** [comment]
> Agreed that generators would be a good fit for this use case, was thinking the lack of support in IE11 would block us using them here though.

We don't need actual syntax support, so as long as we conform to the iterator prototcol it should work.

> Playing around with this, I also wonder how JS Scheduler would know to stop calling the generator.

Can you elaborate? It should stop calling the generator either when the iterator is empty or if it's cancelled.

**Comment by flarnie** [comment]
Haven't finished updating the React callsite but pushed what I have to show the new approach, using the iteration protocol.

**Comment by flarnie** [comment]
Talked to Sophie, and I think it might make sense to just expose 'scheduleLowPriorityWork' and 'scheduleHighPriorityWork', or some variation of that.
Will write an updated proposal and discuss at our meeting tomorrow.

**Comment by acdlite** [comment]
> The problem is we potentially throw errors at the end of the 'performanAsyncWork' call, and that means I can't reliably return the '{value: x, done: y}' object.

I think it's probably fine to schedule the remaining work at the end of the queue if there's an unhandled error.

**Pull request closed without merging**